### PR TITLE
FM-347: add pre-fund to subnet to include genesis balance

### PR DESCRIPTION
**Pull request by adlrocha**
First part of https://github.com/consensus-shipyard/fendermint/issues/347

## Implementation
This PR implements the contract side of https://github.com/consensus-shipyard/fendermint/issues/347.
- It adds a `preFund` method in the subnet actor that allows any address to add some genesis balance into a child subnet before it bootsraps. The reason why we can't hook up this directly to the `fund` method of the gateway is because the gateway has no knowledge of the existence of the subnet until it is bootstrapped and conveniently registered.
- It adds a `genesisCircSupply` argument to the gateway's register method, to allow the registration of subnets with an initial circulating supply. 
- It adds `genesisBalances` and `genesisCircSupply` getters in the subnet actor for Fendermint to be able to pull this information while it builds the child subnet's genesis  

## Next Steps 
- Modify in Fendermint `genesis from-parent` command so it also picks up the genesis balances from the parent and adds them to genesis.
- Add in the `ipc-cli` an option to pre-fund your account when bootstrapping a subnet, and add the corresponding docs.

### CODE REVIEW

**Review by dnkolegov — APPROVED** [review]
Added minor comment.

**Review by cryptoAtwill — APPROVED** [review]

File: src/gateway/GatewayManagerFacet.sol [hunk 1]
```diff
@@ -20,8 +20,11 @@ contract GatewayManagerFacet is GatewayActorModifiers, ReentrancyGuard {
     using SubnetIDHelper for SubnetID;
 
     /// @notice register a subnet in the gateway. It is called by a subnet when it reaches the threshold stake
-    function register() external payable {
-        if (msg.value < s.minStake) {
+    /// The subnet can optionally pass a genesis circulating supply that would be pre-allocated in the
```
[the remark this answers is not in the document]
Comment (adlrocha) [reply src/gateway/GatewayManagerFacet.sol:23]: fixed

File: src/gateway/GatewayManagerFacet.sol [hunk 2]
```diff
@@ -37,6 +40,7 @@ contract GatewayManagerFacet is GatewayActorModifiers, ReentrancyGuard {
         subnet.stake = msg.value;
         subnet.status = Status.Active;
         subnet.genesisEpoch = block.number;
+        subnet.circSupply = genesisCircSupply;
```
[the remark this answers is not in the document]
Comment (adlrocha) [reply src/gateway/GatewayManagerFacet.sol:43]: Good catch, I forgot to change it to collateral.

File: src/subnet/SubnetActorManagerFacet.sol [hunk 3]
```diff
@@ -89,6 +89,26 @@ contract SubnetActorManagerFacet is ISubnetActor, SubnetActorModifiers, Reentran
         }
     }
 
+    /// @notice method to add some initial balance into a subnet that hasn't yet bootstrapped.
+    /// This balance is added to user addresses in genesis, and becomes part of the genesis
```
[the remark this answers is not in the document]
Comment (adlrocha) [reply src/subnet/SubnetActorManagerFacet.sol:93]: fixed

File: src/subnet/SubnetActorManagerFacet.sol [hunk 4]
```diff
@@ -89,6 +89,26 @@ contract SubnetActorManagerFacet is ISubnetActor, SubnetActorModifiers, Reentran
...
+    function preFund() external payable {
+        if (msg.value == 0) {
+            revert NotEnoughFunds();
+        }
+
+        if (s.bootstrapped) {
+            revert SubnetAlreadyBootstrapped();
+        }
+
+        if (s.genesisBalance[msg.sender] == 0) {
```
[the remark this answers is not in the document]
Comment (adlrocha) [reply src/subnet/SubnetActorManagerFacet.sol:104]: Because all users are allowed to pre-fund their account, not only genesis validators.

**Pull request merged**